File: Design-Patterns/GOF-3_Factory_Method/c/notification.c
#include <stdio.h>

/* ---- Product ----
   C nao tem interfaces/classes; emulamos com um "vtable" (struct de ponteiro
   de funcao). Cada Notification concreta so precisa preencher esse ponteiro. */
typedef struct Notification {
    void (*send)(const char *message);
} Notification;

/* ---- ConcreteProducts ---- */
void email_send(const char *message) { printf("[Email] Enviando: %s\n", message); }
void sms_send(const char *message)   { printf("[SMS] Enviando: %s\n", message); }
void push_send(const char *message)  { printf("[Push] Enviando: %s\n", message); }

Notification make_email_notification(void) { Notification n = { email_send }; return n; }
Notification make_sms_notification(void)   { Notification n = { sms_send };   return n; }
Notification make_push_notification(void)  { Notification n = { push_send };  return n; }

/* ---- Creator ----
   Guarda o "factory method" como um ponteiro de funcao que devolve um
   Notification. Cada fabrica concreta aponta para uma make_*_notification. */
typedef struct NotificationFactory {
    Notification (*createNotification)(void);
} NotificationFactory;

void notify(NotificationFactory *factory, const char *message) {
    Notification n = factory->createNotification();  /* factory method */
    n.send(message);
}

int main(void) {
    /* ---- ConcreteCreators ----
       cada entrada do array e uma "fabrica concreta" diferente */
    NotificationFactory factories[] = {
        { make_email_notification },
        { make_sms_notification },
        { make_push_notification },
    };
    int total = (int)(sizeof(factories) / sizeof(factories[0]));

    /* O cliente so conhece NotificationFactory/Notification, nunca as
       funcoes concretas de send/create. */
    for (int i = 0; i < total; i++) {
        notify(&factories[i], "Pedido #1234 confirmado!");
    }
    return 0;
}

File: Design-Patterns/GOF-3_Factory_Method/ptyhon/notification.py
from abc import ABC, abstractmethod


# ---- Product ----
class Notification(ABC):
    @abstractmethod
    def send(self, message: str) -> None:
        ...


# ---- ConcreteProducts ----
class EmailNotification(Notification):
    def send(self, message: str) -> None:
        print(f"[Email] Enviando: {message}")


class SMSNotification(Notification):
    def send(self, message: str) -> None:
        print(f"[SMS] Enviando: {message}")


class PushNotification(Notification):
    def send(self, message: str) -> None:
        print(f"[Push] Enviando: {message}")


# ---- Creator ----
class NotificationFactory(ABC):
    @abstractmethod
    def create_notification(self) -> Notification:
        """Factory Method: cada subclasse decide qual Notification criar."""
        ...

    def notify(self, message: str) -> None:
        """Usa o factory method sem conhecer a classe concreta."""
        notification = self.create_notification()
        notification.send(message)


# ---- ConcreteCreators ----
class EmailNotificationFactory(NotificationFactory):
    def create_notification(self) -> Notification:
        return EmailNotification()


class SMSNotificationFactory(NotificationFactory):
    def create_notification(self) -> Notification:
        return SMSNotification()


class PushNotificationFactory(NotificationFactory):
    def create_notification(self) -> Notification:
        return PushNotification()


def main() -> None:
    factories = [
        EmailNotificationFactory(),
        SMSNotificationFactory(),
        PushNotificationFactory(),
    ]

    # O cliente conhece apenas NotificationFactory / Notification.
    for factory in factories:
        factory.notify("Pedido #1234 confirmado!")


if __name__ == "__main__":
    main()
